benches/trap.rs

```diff
@@ -0,0 +1,146 @@
+use anyhow::Result;
+use criterion::*;
+use wasmtime::*;
+
+criterion_main!(benches);
+criterion_group!(benches, bench_traps);
+
+fn bench_traps(c: &mut Criterion) {
+    bench_multi_threaded_traps(c);
+    bench_many_modules_registered_traps(c);
+    bench_many_stack_frames_traps(c);
+}
+
+fn bench_multi_threaded_traps(c: &mut Criterion) {
+    let mut group = c.benchmark_group("multi-threaded-traps");
+
+    for num_threads in vec![1, 2, 4, 8, 16, 32] {
+        group.throughput(Throughput::Elements(num_threads));
+        group.bench_with_input(
+            BenchmarkId::from_parameter(num_threads),
+            &num_threads,
+            |b, &num_threads| {
+                let engine = Engine::default();
+                let module = module(&engine, 10).unwrap();
+
+                b.iter(|| {
+                    let threads = (0..num_threads)
+                        .map(|_| {
+                            std::thread::spawn({
+                                let engine = engine.clone();
+                                let module = module.clone();
+                                move || {
+                                    const ITERS: u32 = 1000;
+                                    for _ in 0..ITERS {
+                                        let mut store = Store::new(&engine, ());
+                                        let instance =
+                                            Instance::new(&mut store, &module, &[]).unwrap();
+                                        let f = instance
+                                            .get_typed_func::<(), (), _>(&mut store, "")
+                                            .unwrap();
+                                        assert!(f.call(&mut store, ()).is_err());
+                                    }
+                                }
+                            })
+                        })
+                        .collect::<Vec<_>>();
+                    threads.into_iter().for_each(|t| t.join().unwrap());
+                });
+            },
+        );
+    }
```

Comment:
Personally I find that benchmarks like this aren't the best for measuring multithreaded contention. The threads aren't actually guaranteed to be interleaved here as starting up a thread may sometimes take just as long as the number of iterations. Additionally the numbers coming out of criterion itself are unrelated to other numbers below about depth and number of modules since here it's N threads doing 1000 traps where below it's time-per-trap.

Would you be up for restructuring this like the concurrent instantiation benchmark? That one has a scheme where:

* N threads are started in the background infinitely doing work.
* The main thread waits for all N threads to start
* The main thread then does normal criterion-things measuring per-iteration time (e.g. per-one-trap time)
* After the main thread finishes benchmarking it sends a shutdown signal to all threads and joins them

That should guarantee that there's concurrent work in the background and additionally yield comparable numbers to the other benchmarks below.

Reply:
Good idea!

Reply:
Done.